File: Dukan_x/my-backend/src/modules/mobile-shop/config/model-version.config.ts
/**
 * Model Version Configuration
 *
 * Defines supported data-model versions, migration paths, and compatibility windows.
 * Every authoritative record carries a dataModelVersion; readers use pure adapters
 * for earlier supported versions.
 *
 * Requirements: 6.20, 6.33–6.36
 */

export interface ModelVersionConfig {
  /** Current data model version written by the backend */
  readonly currentVersion: number;
  /** Minimum version the backend can read and upgrade transparently */
  readonly minSupportedVersion: number;
  /** Maximum version (equals currentVersion for forward compat) */
  readonly maxSupportedVersion: number;
  /** Minimum API version the backend accepts from clients */
  readonly minSupportedApiVersion: number;
  /** Current API version */
  readonly currentApiVersion: number;
  /** Versions that have explicit migration/backfill support */
  readonly migrationPaths: readonly MigrationPath[];
  /** Version window for queued offline mutations — older payloads are rejected */
  readonly queuedMutationMaxAge: number;
}

export interface MigrationPath {
  /** Source version */
  readonly from: number;
  /** Target version */
  readonly to: number;
  /** Description of what changed */
  readonly description: string;
  /** Whether this migration can run in-place (lazy on read) vs batch backfill */
  readonly lazyUpgrade: boolean;
}

export const MODEL_VERSION_CONFIG: ModelVersionConfig = {
  currentVersion: 1,
  minSupportedVersion: 1,
  maxSupportedVersion: 1,
  minSupportedApiVersion: 1,
  currentApiVersion: 1,

  // As the schema evolves, add migration paths here:
  migrationPaths: [
    // Example for future use:
    // { from: 1, to: 2, description: 'Add reservation expiry field', lazyUpgrade: true },
  ],

  // Queued mutations older than this version delta are rejected on push
  queuedMutationMaxAge: 1, // Must be within 1 version of current
} as const;

File: Dukan_x/lib/features/mobile_shop/config/model_version_config.dart
/// Model Version Configuration (Flutter)
///
/// Defines supported data-model versions and API versions for the client.
/// Used to determine compatibility with backend responses and queued mutations.
///
/// Requirements: 6.20, 6.33–6.36
library;

import 'package:flutter/foundation.dart';

/// Model version configuration.
@immutable
class ModelVersionConfig {
  /// Current data model version the client writes.
  final int currentVersion;

  /// Minimum data model version the client can read.
  final int minSupportedVersion;

  /// Maximum data model version the client supports.
  final int maxSupportedVersion;

  /// Minimum API version the client sends.
  final int minSupportedApiVersion;

  /// Current API version the client sends.
  final int currentApiVersion;

  /// Queued mutations older than this version delta are rejected on push.
  final int queuedMutationMaxAge;

  const ModelVersionConfig({
    required this.currentVersion,
    required this.minSupportedVersion,
    required this.maxSupportedVersion,
    required this.minSupportedApiVersion,
    required this.currentApiVersion,
    required this.queuedMutationMaxAge,
  });

  /// Whether [version] is within the supported read window.
  bool isVersionSupported(int version) {
    return version >= minSupportedVersion && version <= maxSupportedVersion;
  }

  /// Whether a queued mutation at [mutationVersion] is still valid.
  bool isMutationVersionValid(int mutationVersion) {
    return (currentVersion - mutationVersion).abs() <= queuedMutationMaxAge;
  }
}

/// Default model version configuration.
const kModelVersionConfig = ModelVersionConfig(
  currentVersion: 1,
  minSupportedVersion: 1,
  maxSupportedVersion: 1,
  minSupportedApiVersion: 1,
  currentApiVersion: 1,
  queuedMutationMaxAge: 1,
);
